File: client/src/main/java/pro/panopticon/client/sensor/impl/SuccessrateSensor.java
package pro.panopticon.client.sensor.impl;

import com.amazonaws.services.cloudwatch.model.StandardUnit;
import org.apache.commons.collections4.queue.CircularFifoQueue;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import pro.panopticon.client.model.Measurement;
import pro.panopticon.client.sensor.Sensor;
import pro.panopticon.client.util.NowSupplier;
import pro.panopticon.client.util.NowSupplierImpl;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;

import static java.util.stream.Collectors.toList;

public class SuccessrateSensor implements Sensor {

    public static final int HOURS_FOR_ERROR_TICK_TO_BE_CONSIDERED_OUTDATED = 1;
    private final Logger LOG = LoggerFactory.getLogger(this.getClass());

    /**
     * Number of events to keep.
     */
    private final int numberToKeep;

    /**
     * Triggers an alert to Slack when reached.
     * Should always be a Double between 0.00 and 1.00
     * Format: percentage / 100
     * <p>
     * Example: 0.1 will trigger a warning at 10% failure rate
     */
    private final Double warnLimit;

    /**
     * Triggers an alert to Slack and PagerDuty when reached.
     * Should always be a Double between 0.00 and 1.00
     * Format: percentage / 100
     * <p>
     * Example: 0.2 will trigger an alert at 20% failure rate
     */
    private final Double errorLimit;

    private final Map<AlertInfo, CircularFifoQueue<Tick>> eventQueues = new HashMap<>();
    private final NowSupplier nowSupplier;

    public SuccessrateSensor(int numberToKeep, Double warnLimit, Double errorLimit) {
        this.numberToKeep = numberToKeep;
        this.warnLimit = warnLimit;
        this.errorLimit = errorLimit;
        this.nowSupplier = new NowSupplierImpl();
    }

    SuccessrateSensor(int numberToKeep, Double warnLimit, Double errorLimit, NowSupplier nowSupplier) {
        this.numberToKeep = numberToKeep;
        this.warnLimit = warnLimit;
        this.errorLimit = errorLimit;
        this.nowSupplier = nowSupplier;
    }
    @Deprecated
    public synchronized void tickSuccess(String key) {
        tickSuccess(new AlertInfo(key, null));
    }

    @Deprecated
    public synchronized void tickFailure(String key) {
        tickFailure(new AlertInfo(key, null));
    }

    public synchronized void tickFailure(AlertInfo alertInfo) {
        try {
            getQueueForKey(alertInfo).add(new Tick(Event.FAILURE, nowSupplier.now()));
        } catch (Exception e) {
            LOG.warn("Something went wrong when counting FAILURE for " + alertInfo.getSensorKey(), e);
        }
    }

    public synchronized void tickSuccess(AlertInfo alertInfo) {
        try {
            getQueueForKey(alertInfo).add(new Tick(Event.SUCCESS, nowSupplier.now()));
        } catch (Exception e) {
            LOG.warn("Something went wrong when counting SUCCESS for " + alertInfo.getSensorKey(), e);
        }
    }

    private CircularFifoQueue<Tick> getQueueForKey(AlertInfo key) {
        return eventQueues.computeIfAbsent(key, k -> new CircularFifoQueue<>(numberToKeep));
    }

    @Override
    public List<Measurement> measure() {
        return eventQueues.entrySet().stream()
                .map(this::measure)
                .collect(toList());
    }

    private Measurement measure(Map.Entry<AlertInfo, CircularFifoQueue<Tick>> e) {
        AlertInfo alertInfo = e.getKey();
        List<Tick> ticks = new ArrayList<>(e.getValue());
        int all = ticks.size();
        long success = ticks.stream().filter(tick -> tick.event == Event.SUCCESS).count();
        long failure = ticks.stream().filter(tick -> tick.event == Event.FAILURE).count();
        double percentFailureDouble = all > 0 ? (double) failure / (double) all : 0;
        boolean enoughDataToAlert = all == numberToKeep;
        boolean hasRecentErrorTicks = hasRecentErrorTicks(ticks);
        String display = String.format("Last %s calls: %s success, %s failure (%.2f%% failure)%s%s",
                Integer.min(all, numberToKeep),
                success,
                all - success,
                percentFailureDouble * 100,
                enoughDataToAlert ? "" : " - not enough calls to report status yet",
                hasRecentErrorTicks ? "" : " - no recent error ticks"
        );
        String status = decideStatus(enoughDataToAlert, percentFailureDouble, hasRecentErrorTicks);
        return new Measurement(alertInfo.getSensorKey(), status, display, new Measurement.CloudwatchValue(percentFailureDouble * 100, StandardUnit.Percent), alertInfo.getDescription());
    }

    private String decideStatus(boolean enoughDataToAlert, double percentFailure, boolean hasRecentErrorTicks) {
        if (!enoughDataToAlert) return "INFO";
        if (!hasRecentErrorTicks) return "INFO";
        if (errorLimit != null && percentFailure >= errorLimit) return "ERROR";
        if (warnLimit != null && percentFailure >= warnLimit) return "WARN";
        return "INFO";
    }

    private boolean hasRecentErrorTicks(List<Tick> events) {
        return events.stream()
                .filter(tick -> tick.event == Event.FAILURE)
                .anyMatch(tick -> ChronoUnit.HOURS.between(tick.createdAt, nowSupplier.now()) < HOURS_FOR_ERROR_TICK_TO_BE_CONSIDERED_OUTDATED);
    }

    private enum Event {
        SUCCESS,
        FAILURE
    }

    private static class Tick {
        private final Event event;
        private final LocalDateTime createdAt;

        private Tick(Event event, LocalDateTime createdAt) {
            this.event = event;
            this.createdAt = createdAt;
        }
    }

}

File: client/src/main/java/pro/panopticon/client/sensor/impl/SuccessrateSensor.kt
package pro.panopticon.client.sensor.impl

import com.amazonaws.services.cloudwatch.model.StandardUnit
import mu.KotlinLogging
import org.apache.commons.collections4.queue.CircularFifoQueue
import pro.panopticon.client.model.Measurement
import pro.panopticon.client.sensor.Sensor
import pro.panopticon.client.util.NowSupplier
import pro.panopticon.client.util.NowSupplierImpl
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import kotlin.math.min

open class SuccessrateSensor : Sensor {
    private val LOG = KotlinLogging.logger { }

    /**
     * Number of events to keep.
     */
    private val numberToKeep: Int

    /**
     * Triggers an alert to Slack when reached.
     * Should always be a Double between 0.00 and 1.00
     * Format: percentage / 100
     *
     *
     * Example: 0.1 will trigger a warning at 10% failure rate
     */
    private val warnLimit: Double?

    /**
     * Triggers an alert to Slack and PagerDuty when reached.
     * Should always be a Double between 0.00 and 1.00
     * Format: percentage / 100
     *
     *
     * Example: 0.2 will trigger an alert at 20% failure rate
     */
    private val errorLimit: Double?
    private val eventQueues: MutableMap<Sensor.AlertInfo, CircularFifoQueue<Tick>> = mutableMapOf()
    private val previousErrorPercentage = mutableMapOf<Sensor.AlertInfo, Double>()
    private val nowSupplier: NowSupplier

    constructor(numberToKeep: Int, warnLimit: Double?, errorLimit: Double?) {
        this.numberToKeep = numberToKeep
        this.warnLimit = warnLimit
        this.errorLimit = errorLimit
        nowSupplier = NowSupplierImpl()
    }

    internal constructor(numberToKeep: Int, warnLimit: Double?, errorLimit: Double?, nowSupplier: NowSupplier) {
        this.numberToKeep = numberToKeep
        this.warnLimit = warnLimit
        this.errorLimit = errorLimit
        this.nowSupplier = nowSupplier
    }

    @Synchronized
    fun tickFailure(alertInfo: Sensor.AlertInfo) {
        try {
            getQueueForKey(alertInfo).add(Tick(Event.FAILURE, nowSupplier.now()))
        } catch (e: Exception) {
            LOG.warn("Something went wrong when counting FAILURE for " + alertInfo.sensorKey, e)
        }
    }

    @Synchronized
    fun tickAndLogFailure(alertInfo: Sensor.AlertInfo, vararg logAppends: String?) {
        tickFailure(alertInfo)
        LOG.info("SENSOR SUCCESS RATE - [FAILURE] - [" + alertInfo.sensorKey + "] - [" + alertInfo.description + "] - " + logAppends.joinToString(
            " - ") { "[$it]" })
    }

    @Synchronized
    fun tickSuccess(alertInfo: Sensor.AlertInfo) {
        try {
            getQueueForKey(alertInfo).add(Tick(Event.SUCCESS, nowSupplier.now()))
        } catch (e: Exception) {
            LOG.warn("Something went wrong when counting SUCCESS for " + alertInfo.sensorKey, e)
        }
    }

    @Synchronized
    fun tickAndLogSuccess(alertInfo: Sensor.AlertInfo, vararg logAppends: String?) {
        tickSuccess(alertInfo)
        LOG.info("SENSOR SUCCESS RATE - [SUCCESS] - [" + alertInfo.sensorKey + "] - [" + alertInfo.description + "] - " + logAppends.joinToString(
            " - ") { "[$it]" })
    }

    private fun getQueueForKey(key: Sensor.AlertInfo): CircularFifoQueue<Tick> {
        return eventQueues.computeIfAbsent(key) { CircularFifoQueue(numberToKeep) }
    }

    override fun measure(): List<Measurement> {
        return eventQueues.entries
            .map { this.measure(it.key, it.value) }
    }

    private fun measure(alertInfo: Sensor.AlertInfo, ticks: CircularFifoQueue<Tick>): Measurement {
        val all = ticks.size
        val success = ticks.count { tick: Tick -> tick.event == Event.SUCCESS }
        val failure = ticks.count { tick: Tick -> tick.event == Event.FAILURE }
        val previousFailureDouble = previousErrorPercentage[alertInfo] ?: 0.0
        val percentFailureDouble: Double = (if (all > 0) failure.toDouble() / all.toDouble() else 0.0)
            .also { previousErrorPercentage[alertInfo] = it }
        val enoughDataToAlert = all == numberToKeep
        val hasRecentErrorTicks = hasRecentErrorTicks(ticks)
        val display = String.format("Last %s calls: %s success, %s failure (%.2f%% failure)%s%s",
            min(all, numberToKeep),
            success,
            all - success,
            percentFailureDouble * 100,
            if (enoughDataToAlert) "" else " - not enough calls to report status yet",
            if (hasRecentErrorTicks) "" else " - no recent error ticks",
            when {
                previousFailureDouble == percentFailureDouble -> ""
                previousFailureDouble < percentFailureDouble -> " :chart_with_upwards_trend:"
                else -> " :chart_with_downwards_trend:"
            }
        )
        val status = decideStatus(enoughDataToAlert, percentFailureDouble, hasRecentErrorTicks)
        return Measurement(
            key = alertInfo.sensorKey,
            status = status,
            cloudwatchValue = Measurement.CloudwatchValue(percentFailureDouble * 100, StandardUnit.Percent),
            displayValue = display,
            description = alertInfo.description,
        )
    }

    private fun decideStatus(
        enoughDataToAlert: Boolean,
        percentFailure: Double,
        hasRecentErrorTicks: Boolean,
    ): Measurement.Status {
        return when {
            !enoughDataToAlert -> Measurement.Status.INFO
            !hasRecentErrorTicks -> Measurement.Status.INFO
            (errorLimit != null && percentFailure >= errorLimit) -> Measurement.Status.ERROR
            (warnLimit != null && percentFailure >= warnLimit) -> Measurement.Status.WARN
            else -> Measurement.Status.INFO
        }
    }

    private fun hasRecentErrorTicks(ticks: CircularFifoQueue<Tick>): Boolean {
        return ticks
            .filter { tick: Tick -> tick.event == Event.FAILURE }
            .any { tick: Tick ->
                ChronoUnit.HOURS.between(tick.createdAt,
                    nowSupplier.now()) < HOURS_FOR_ERROR_TICK_TO_BE_CONSIDERED_OUTDATED
            }
    }

    private enum class Event {
        SUCCESS, FAILURE
    }

    private class Tick(val event: Event, val createdAt: LocalDateTime)
    companion object {
        const val HOURS_FOR_ERROR_TICK_TO_BE_CONSIDERED_OUTDATED = 1
    }
}
